extensions/HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -13,83 +13,106 @@ namespace HyperVExtension.Services;
 
     public VMGalleryService(IDownloaderService downloaderService)
     {
         _downloaderService = downloaderService;
     }
 
-    private static readonly Uri VmGalleryUrl = new("https://go.microsoft.com/fwlink/?linkid=851584");
+    private static readonly Uri _vmGalleryUrl = new("https://go.microsoft.com/fwlink/?linkid=851584");
 
-    private VMGalleryImageList _imageList = new();
+    private readonly VMGalleryImageList _imageList = new();
 
     /// <inheritdoc cref="IVMGalleryService.GetGalleryImagesAsync"/>
     public async Task<VMGalleryImageList> GetGalleryImagesAsync()
-    {
-        // If we have already downloaded the image list before, return it.
-        if (_imageList.Images.Count > 0)
-        {
-            return _imageList;
-        }
-
-        var emptyList = new VMGalleryImageList();
-        try
-        {
-            var cancellationTokenSource = new CancellationTokenSource();
-
-            // This should be quick as the file is around 28.0 KB, so a 5 minute timeout should be ok on the worst network.
-            cancellationTokenSource.CancelAfter(TimeSpan.FromMinutes(5));
-
-            // get the JSON data
-            var resultJson = await _downloaderService.DownloadStringAsync(VmGalleryUrl.AbsoluteUri, cancellationTokenSource.Token);
-            _imageList = JsonSerializer.Deserialize(resultJson, typeof(VMGalleryImageList), _jsonOptions) as VMGalleryImageList ?? emptyList;
-
-            // Now we need to download the base64 images for the symbols (icons). So they can be used within an adaptive card.
-            foreach (var image in _imageList.Images)
-            {
-                if (!string.IsNullOrEmpty(image.Symbol.Uri))
-                {
-                    var byteArray = await _downloaderService.DownloadByteArrayAsync(image.Symbol.Uri, cancellationTokenSource.Token);
-                    if (!ValidateFileSha256Hash(byteArray, image.Symbol.Hash))
-                    {
-                        _log.Error($"Symbol Hash '{image.Symbol.Hash}' validation failed for image with name '{image}'. Symbol uri: '{image.Symbol}'");
-                        continue;
-                    }
-
-                    image.Symbol.Base64Image = Convert.ToBase64String(byteArray);
-                }
-
-                if (!string.IsNullOrEmpty(image.Disk.Uri))
-                {
-                    var totalSizeOfDisk = await _downloaderService.GetHeaderContentLength(new Uri(image.Disk.Uri), cancellationTokenSource.Token);
-                    if (ulong.TryParse(image.Requirements.DiskSpace, CultureInfo.InvariantCulture, out var requiredDiskSpace))
-                    {
-                        image.Disk.SizeInBytes = (ulong)totalSizeOfDisk;
-                    }
-                }
-            }
+    {
+        try
+        {
+            await _gallerySemaphore.WaitAsync();
+
+            // If we have already downloaded the image list before, return it.

```

Comment:
Might be better to have this early return before the semaphore wait.

Reply:
if we return early we might end up in a situation where one caller gets a list with 1 vm image and another gets a full list because the `_imageList` variable would be accessed on different threads.